openmeter/streaming/clickhouse/cache.go

```diff
@@ -0,0 +1,276 @@
+		if lastCachedWindow.Equal(*cacheableQueryMeter.To) {
+			c.config.Logger.Debug("no new rows to query for cache period, returning cached data", "count", len(values))
+
+			return createRemainingQuery(cacheableQueryMeter), values, nil
+		}
+	}
+
+	// Step 2: Query new rows for the uncached time period
+	newRows, err := c.queryMeter(ctx, cacheableQueryMeter)
+	if err != nil {
+		return originalQueryMeter, values, fmt.Errorf("query new meter rows: %w", err)
+	}
+
+	values = append(values, newRows...)
+
+	// Step 3: Cache the new results
+	if len(newRows) > 0 {
+		if err := c.storeCachedMeterRows(ctx, hash, cacheableQueryMeter, newRows); err != nil {
+			// Log the error but don't fail the query
+			c.config.Logger.Error("failed to store new rows in cache", "error", err, "from", cacheableQueryMeter.From, "to", cacheableQueryMeter.To, "count", len(newRows))
+		} else {
+			c.config.Logger.Debug("new rows stored in cache", "from", cacheableQueryMeter.From, "to", cacheableQueryMeter.To, "count", len(newRows))
+		}
+	}
+
+	// Result
+	c.config.Logger.Debug("returning cached and new rows", "from", cacheableQueryMeter.From, "to", cacheableQueryMeter.To, "count", len(values))
+
+	return createRemainingQuery(cacheableQueryMeter), values, nil
+}
+
+// prepareCacheableQueryPeriod prepares the time range for cacheable queries
+func (c *Connector) prepareCacheableQueryPeriod(originalQueryMeter queryMeter) (queryMeter, error) {
+	cacheableQueryMeter := originalQueryMeter
+	now := time.Now().UTC()
+
+	if originalQueryMeter.From == nil {
+		return cacheableQueryMeter, fmt.Errorf("from is required for cached queries")
+	}
+
+	// Set the end time to now if not provided
+	if cacheableQueryMeter.To == nil {
+		cacheableQueryMeter.To = &now
+	}
+
+	// We do not cache data that is less than 24 hours old
+	toFreshness := now.Sub(*cacheableQueryMeter.To)
+
+	if toFreshness < c.config.QueryCacheMinimumCacheableUsageAge {
+		delta := c.config.QueryCacheMinimumCacheableUsageAge - toFreshness
+
+		cacheableQueryMeter.To = lo.ToPtr(cacheableQueryMeter.To.Add(-delta))
+	}
+
+	// We truncate to complete days to avoid partial days in the cache
+	cacheableQueryMeter.To = lo.ToPtr(cacheableQueryMeter.To.Truncate(time.Hour * 24))
+
+	// This is the window size that the cache will use if no window size is provided
+	if cacheableQueryMeter.WindowSize == nil {
+		cacheableQueryMeter.WindowSize = lo.ToPtr(meterpkg.WindowSizeDay)
+	}
+
+	return cacheableQueryMeter, nil
+}
+
+// fetchCachedMeterRows queries the meter_query_hash table for cached results
+func (c *Connector) fetchCachedMeterRows(ctx context.Context, hash string, queryMeter queryMeter) ([]meterpkg.MeterQueryRow, error) {
+	var cachedValues []meterpkg.MeterQueryRow
+
+	cacheQuery := getMeterQueryRowsFromCache{
+		Database:  queryMeter.Database,
+		TableName: meterQueryRowCacheTable,
+		Hash:      hash,
+		Namespace: queryMeter.Namespace,
+		From:      queryMeter.From,
+		To:        queryMeter.To,
```

Comment:
I would mandate To to be not nil (maybe from too, but that's less problematic), so that we are not storing any cache item that's "open ended"

Reply:
We discussed it separately. With open-ended `to` we don't cache the last 3 days (default, see `QueryCacheMinimumCacheableUsageAge`)
See: https://github.com/openmeterio/openmeter/pull/2392/files#diff-42194278b2abea7e30a5260024bfbac0a8c925e4fa207230bb2d31016a22cfafR133-R157